libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/analysis/ImmutableValidator.h>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace solidity::frontend;
+
+void ImmutableValidator::analyze()
+{
+	for (VariableDeclaration const* stateVar: m_currentContract->stateVariablesIncludingInherited())
+		if (stateVar->immutable() && stateVar->value())
```

Comment:
We also have to visit all non-constant state variables. They value is evaluated at compile-time and can call arbitrary functions.